### PR TITLE
RCHAIN-3365: Use per-validator vaults in PoS

### PR DESCRIPTION
**Pull request by tymm**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/BlockCreator.scala [hunk 1]
```diff
@@ -193,6 +194,7 @@ object BlockCreator {
       deploys: Seq[DeployData],
       justifications: Seq[Justification],
       maxBlockNumber: Long,
+      sender: PublicKey,
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/BlockCreator.scala:204]: I feel that this should be named "creator" instead of sender?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -1405,6 +1428,7 @@ class DebruijnInterpreter[M[_], F[_]](
       "size"        -> size,
       "length"      -> length,
       "slice"       -> slice,
+      "take"        -> take,
```
Comment (odeac) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1430]: I'm not sure this should be a method or a contract in ListOps.rho. My feeling is that, since map is a contract in ListOps.rho then take should be too

File: casper/src/test/resources/RevAddressTest.rho [hunk 3]
```diff
@@ -79,7 +79,9 @@ in {
     new unf1, unf2, r1, r2 in {
       revAddress!("fromUnforgeable", *unf1, *r1) |
       revAddress!("fromUnforgeable", *unf2, *r2) |
-      rhoSpec!("assert", (*r1, "!= <-", *r2), "expecting RevAddresses to be unequal", *ackCh)
+      for (@fromUnf1 <- r1) {
+        rhoSpec!("assert", (fromUnf1, "!= <-", *r2), "expecting RevAddresses to be unequal", *ackCh)
```
Comment (odeac) [inline casper/src/test/resources/RevAddressTest.rho:87]: I would also add a positive test like deriving twice from the same unforgeable name returns the same address

File: casper/src/test/resources/PoSTest.rho [hunk 4]
```diff
@@ -97,15 +97,23 @@ match (
...
-                  [
-                    (true, "all good here")
-                  ],
-                  *ackCh
-                )
+                // After the first "closeBlock" is done, there are active validators
+                // For testing we need one of the active validators to be the block sender
+                @PoS!("getActiveValidators", *activeValidatorsCh) |
+                for (@activeValidators <- activeValidatorsCh) {
+                  setBlockData!("sender", activeValidators.nth(0), *blockDataSet) |
```
Comment (odeac) [inline casper/src/test/resources/PoSTest.rho]: why do you add this assertion to "test_close_block_finishes"? Are they related?

I think a better place would be to add this assert to the test of getBlockData

File: casper/src/main/resources/PoS.rhox [hunk 5]
```diff
@@ -429,13 +429,15 @@ new  PoS,
...
+        new availableValidatorsCh, isAvailable, pksCh in {
           @ListOps!("filter", allBonds.toList(), *isAvailable, *availableValidatorsCh) |
           contract isAvailable(@(pk, _), resultCh) = { resultCh!(not withdrawers.contains(pk))} |
 
           for (@availableValidators <- availableValidatorsCh) {
-            //TODO choose only a limited number of validators
-            @ListOps!("map", availableValidators, *fst, *returnCh)
+            @ListOps!("map", availableValidators, *fst, *pksCh) |
+            for (@pks <- pksCh) {
+              returnCh!(pks.take($$numberOfActiveValidators$$))
```
Comment (odeac) [inline casper/src/main/resources/PoS.rhox:515]: This is not a fair random selection. I think we should allow all the validators be active in the first epoch and then, in the second epoch, when we have a random number produced from all the validators, we can fairly select 100 of them

File: casper/src/main/resources/PoS.rhox [hunk 6]
```diff
@@ -45,6 +46,25 @@ new  PoS,
...
+              for (@(true, vault) <- vaultCh) {
+                returnCh!((*unf, revAddress, vault))
+              }
+            }
+          }
+        } |
+
+        @ListOps!("range", 0, $$numberOfActiveValidators$$, *rangeCh) |
+        for (@range <- rangeCh) {
+          @ListOps!("map", range, *makeVault, *perActiveValidatorPosVaultsCh)
```
Comment (odeac) [inline casper/src/main/resources/PoS.rhox]: you can use "parMap" here

File: casper/src/main/resources/PoS.rhox [hunk 7]
```diff
@@ -344,17 +405,25 @@ new  PoS,
...
               authKeyCh in {
             @RevVault!("deployerAuthKey", *authKeyCh) |
+            getBlockData!(*blockDataCh) |
             for (@authKey <- authKeyCh) {
               revAddressOps!("fromPublicKey", fromUserPk, *revAddressCh) |
               for (@fromRevAddress <- revAddressCh) {
                 @RevVault!("findOrCreate", fromRevAddress, *vaultCh) |
-                for (@(true, fromVault) <- vaultCh) {
-                  @fromVault!("transfer", posRevAddress, amount, authKey, *returnCh)
+                for (@(true, fromVault) <- vaultCh; _, _, @sender <- blockDataCh) {
```
Comment (odeac) [inline casper/src/main/resources/PoS.rhox]: I would put the receive from blockDataCh closer to the getBlockData!(*blockDataCh) for better readability. One next to the other is the easier to match visually

File: casper/src/main/scala/coop/rchain/casper/genesis/Genesis.scala [hunk 8]
```diff
@@ -13,17 +13,25 @@ import coop.rchain.casper.util.Sorting.byteArrayOrdering
...
 import coop.rchain.shared.{Log, LogSource, Time}
 
 final case class Genesis(
     shardId: String,
     timestamp: Long,
     proofOfStake: ProofOfStake,
     vaults: Seq[Vault],
-    supply: Long
+    supply: Long,
+    publicKey: PublicKey
```
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/genesis/Genesis.scala]: It looks like this field is only accessed to do the assignment `publicKey = Genesis.pk`. It also looks like the only reason we need this field in the first place is so that a valid key is returned in `getBlockData`. Since this data is going to be over-written upon the first block, and (if I'm not mistaken) we don't need it anywhere outside of Rholang. So, could we just `Secp256k1.newKeyPair._2` in `computeGenesis` in `RuntimeManager`?

File: casper/src/main/resources/PoS.rhox [hunk 9]
```diff
@@ -344,17 +405,25 @@ new  PoS,
...
+                  @ListOps!("indexOf", activeValidators, sender, *indexSender) |
+                  for (@idx <- indexSender) {
+                    for (@vaults <<- perActiveValidatorPosVaultsCh) {
+                      @fromVault!("transfer", vaults.nth(idx).nth(1), amount, authKey, *returnCh)
```
Comment (odeac) [inline casper/src/main/resources/PoS.rhox]: Can you explain why is this needed "vaults.nth(idx).nth(1)"?

**Review by odeac — CHANGES_REQUESTED** [review]

File: casper/src/test/resources/PoSTest.rho [hunk 10]
```diff
@@ -97,15 +97,23 @@ match (
...
               for ( _ <- resultCh) {
-                rhoSpec!("assertMany",
-                  [
-                    (true, "all good here")
-                  ],
-                  *ackCh
-                )
+                // After the first "closeBlock" is done, there are active validators
+                // For testing we need one of the active validators to be the block sender
+                @PoS!("getActiveValidators", *activeValidatorsCh) |
```
Comment (JosephDenman) [inline casper/src/test/resources/PoSTest.rho]: This method isn't added until `8199c63e`. This should probably be introduced after or with that commit.

**Review by JosephDenman — CHANGES_REQUESTED** [review]
Still trying to grok the updated PoS logic. Will continue the review tomorrow.

File: casper/src/main/scala/coop/rchain/casper/BlockCreator.scala (hunk 1 above)
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/BlockCreator.scala:204]: Since this directly relates to the `sender` field in `BlockMessage`, I feel like `sender` is the better name. Happy to change it if @JosDenmark also prefers "creator" :)

File: casper/src/test/resources/PoSTest.rho (hunk 4 above)
Comment (tymm) [reply casper/src/test/resources/PoSTest.rho]: For the other tests to work, we have to set the block sender to one of the active validators.
Since we potentially pick a new set of active validators in `PoS!closeBlock`, we have to set the sender after running `PoS!closeBlock`.

I just realized that we should do that after each call to `PoS!closeBlock`. So I guess I should create a helper contract that does that and replace every occurrence of `@PoS!("closeBlock", *closeBlockCh)` in `PoSTest.rho` with the helper contract.

File: casper/src/main/scala/coop/rchain/casper/BlockCreator.scala (hunk 1 above)
Comment (JosephDenman) [reply casper/src/main/scala/coop/rchain/casper/BlockCreator.scala:204]: I'm indifferent.

File: casper/src/main/resources/PoS.rhox [hunk 11]
```diff
@@ -344,17 +405,25 @@ new  PoS,
...
 
-        contract deposit (@fromUserPk, @amount, returnCh) = {
+        contract deposit (@fromUserPk, @amount, @activeValidators, returnCh) = {
           new vaultCh,
               revAddressCh,
+              indexSender,
+              getBlockData(`rho:block:data`), blockDataCh,
               authKeyCh in {
             @RevVault!("deployerAuthKey", *authKeyCh) |
+            getBlockData!(*blockDataCh) |
```
Comment (JosephDenman) [inline casper/src/main/resources/PoS.rhox]: This code would be more readable if the send on `getBlockData` was in the same lexical scope as the receive on `getBlockData`.

File: casper/src/main/resources/PoS.rhox [hunk 12]
```diff
@@ -344,17 +405,25 @@ new  PoS,
...
         } |
 
-        contract deposit (@fromUserPk, @amount, returnCh) = {
+        contract deposit (@fromUserPk, @amount, @activeValidators, returnCh) = {
           new vaultCh,
               revAddressCh,
+              indexSender,
+              getBlockData(`rho:block:data`), blockDataCh,
               authKeyCh in {
             @RevVault!("deployerAuthKey", *authKeyCh) |
```
Comment (JosephDenman) [inline casper/src/main/resources/PoS.rhox:418]: I think this should be run concurrently with `revAddressOps!("fromPublicKey", ...)`. Not only does it increase concurrency, it also makes data dependencies syntactically evident. You should also review the Rholang you added to ensure that this is generally true. Same with the lexical scope comment below.

**Review by JosephDenman — COMMENTED** [review]
I know this isn't part of this PR, but while you're at it, would you change `$$quarantineLength$$ + $$epochLength$$ * (1 + blockNumber / $$epochLength$$)))` in `PoS` to `$$quarantineLength$$ + $$epochLength$$  + blockNumber`? It's also not immediately obvious to me why this is the block number at which a validator should be able to cash out. @odeac Please shed some light so the description can be included in a comment.

File: casper/src/main/resources/PoS.rhox (hunk 9 above)
Comment (tymm) [reply casper/src/main/resources/PoS.rhox]: I added a `match` to make it a little easier to understand.

`vaults.nth(idx)` is a triple consisting of an unforgeable name, a rev address derived from that unforgeable name and a vault instance derived from the rev address.
`vaults.nth(idx).nth(1)` is therefore needed to get the rev address of the vault that is associated to the current block sender.

File: casper/src/main/resources/PoS.rhox [hunk 13]
```diff
@@ -344,17 +411,25 @@ new  PoS,
...
+            getBlockData!(*blockDataCh) |
             for (@authKey <- authKeyCh) {
               revAddressOps!("fromPublicKey", fromUserPk, *revAddressCh) |
               for (@fromRevAddress <- revAddressCh) {
                 @RevVault!("findOrCreate", fromRevAddress, *vaultCh) |
-                for (@(true, fromVault) <- vaultCh) {
-                  @fromVault!("transfer", posRevAddress, amount, authKey, *returnCh)
+                for (@(true, fromVault) <- vaultCh; _, _, @sender <- blockDataCh) {
```
Comment (lukasz-golebiewski) [inline casper/src/main/resources/PoS.rhox]: Joined channels are more prone to cause conflicts because the mergeability rules are not applied to them. 
I'm not sure if it makes sense to get rid of this particular join, just double-checking.

Comment (tymm) [reply casper/src/main/resources/PoS.rhox]: Hmm, why "more prone"? In which cases are they not conflicting?

Comment (lukasz-golebiewski) [reply casper/src/main/resources/PoS.rhox]: When disparate channels are used in both blocks

File: casper/src/main/scala/coop/rchain/casper/BlockCreator.scala (hunk 1 above)
Comment (odeac) [reply casper/src/main/scala/coop/rchain/casper/BlockCreator.scala:204]: Sender makes sense in the context of message routing. When discussing blocks in the DAG, for me "sender" doesn't make sense anymore. Anyway, I don't have strong opinions about this

File: casper/src/main/scala/coop/rchain/casper/genesis/Genesis.scala (hunk 8 above)
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/genesis/Genesis.scala]: `Secp256k1.newKeyPair._2` doesn't work out since during replay we will get another value. And in general we want to keep things referentially transparent. Instead I changed it to `PublicKey(Array[Byte]())`.

File: casper/src/main/resources/PoS.rhox (hunk 13 above)
Comment (tymm) [reply casper/src/main/resources/PoS.rhox]: Then this join should not be conflicting since new channels get created with every contract call.

**Comment by tymm** [comment]
bors try

**Comment by tymm** [comment]
@odeac and @JosDenmark please have another look. If you prefer the squashed commit log before approving, let me know.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1430]: It looks like the performance hit of using a ListOps version of take is so big that integration tests fail due to timeouts. Also I had to increase the timeout for the `PoSSpec`.

**Comment by JosephDenman** [comment]
@tymm The squashed commit log would be much easier to review. Thanks.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1430]: Does this mean that the Rholang implementation of ListOps is not really usable?

Do we have a rule about putting list operations in the list methods or in the ListOps contract?

**Comment by tymm** [comment]
@JosDenmark done.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (tymm) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1430]: Let me take that back. Actually integration tests failing because of timeouts is probably also due to the overall performance decrease that comes with the new PoS logic.
However `PoSSpec` passed on my laptop with a 120s timeout before the change to `ListOps!take`. Which wasn't the case afterwards.

There was a discussion between Joseph, Artur and Mike in #rholang and I think everyone agreed that methods are not the way to go.

Comment (JosephDenman) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1430]: @tymm That was a slightly different conversation. It was a debate to determine whether the cryptographic functions should be methods or system processes. I would go with adding `take` as a method alongside `slice` and `nth`. It's bound to be more performant and won't introduce unnecessary bloat in the tuplespace. It's also syntactically cleaner.

File: casper/src/main/resources/PoS.rhox (hunk 5 above)
Comment (odeac) [reply casper/src/main/resources/PoS.rhox:515]: do you really need the take here? Can't you just use all the pks as active validators?

File: casper/src/main/resources/PoS.rhox [hunk 14]
```diff
@@ -344,17 +405,29 @@ new  PoS,
...
+                  @ListOps!("indexOf", activeValidators, sender, *indexSender) |
+                  for (@idx <- indexSender) {
+                    for (@vaults <<- perActiveValidatorPosVaultsCh) {
+                      match vaults.nth(idx).nth(1) {
+                        activeValidatorVault => {
+                          @fromVault!("transfer", activeValidatorVault, amount, authKey, *returnCh)
+                        }
+                      }
```
Comment (odeac) [inline casper/src/main/resources/PoS.rhox:430]: this looks better as it makes it obvious that the second nth is applied on a tuple

```scala
                     match vaults.nth(idx) {
                        (activeValidatorVault, _) => {
                          @fromVault!("transfer", activeValidatorVault, amount, authKey, *returnCh)
                        }
                      }
```

File: casper/src/test/resources/PoSTest.rho [hunk 15]
```diff
@@ -463,6 +465,42 @@ match (
...
+              }
+            }
+          } |
+
+          contract closeBlock(ackCh) = {
+            new ackCh0, activeValidatorsCh, setBlockData(`rho:test:block:data:set`), blockDataSet in {
+              @PoS!("closeBlock", *ackCh0) |
+              @PoS!("getActiveValidators", *activeValidatorsCh) |
+              for (@ack <- ackCh0; @activeValidators <- activeValidatorsCh) {
+                // After the first "closeBlock" is done, there are active validators
```
Comment (odeac) [inline casper/src/test/resources/PoSTest.rho]: I don't understand this comment. There should be active validators all the time, not only after the first closeBlock.

File: casper/src/main/resources/PoS.rhox [hunk 16]
```diff
@@ -77,6 +96,21 @@ new  PoS,
...
+            contract PoS (@"getActiveValidators", returnCh) = {
+              new tmpCh in {
+                getMVar!(*stateCh, *tmpCh) |
+                for (@state <- tmpCh) {
+                  returnCh!(state.get("activeValidators"))
+                }
+              }
+            } |
+
+            contract PoS (@"getActiveValidatorVaults", returnCh) = {
```
Comment (odeac) [inline casper/src/main/resources/PoS.rhox:108]: I don't think this should be part of the PoS API. The fact that we have a vault per validator should be only an implementation detail

File: casper/src/test/resources/PoSTest.rho [hunk 17]
```diff
@@ -467,6 +470,42 @@ match (
...
+
+              @posVault!("balance", *posVaultBalanceCh) |
+              for (@posVaultBalance <- posVaultBalanceCh) {
+                @PoS!("getActiveValidatorVaults", *vaultsCh) |
+                for (@vaults <- vaultsCh) {
+                  @ListOps!("fold", vaults, posVaultBalance, *accBalances, *retCh)
+                }
+              }
+            }
+          } |
```
Comment (odeac) [inline casper/src/test/resources/PoSTest.rho:494]: I think that the logic of the getPoSBalance contract should be part of the PoS API

**Review by odeac — CHANGES_REQUESTED** [review]
I still have a problem with the fact that the vault-per-validator implementation is visible in the API. Do we really need that?

File: casper/src/test/scala/coop/rchain/casper/helper/BlockDataContract.scala [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+  def set[F[_]: Concurrent: Span](
+      ctx: SystemProcess.Context[F]
+  )(message: (Seq[ListParWithRandom], Int)): F[Unit] = {
+
+    val isContractCall = new ContractCall(ctx.space, ctx.dispatcher)
+    message match {
+      case isContractCall(
+          produce,
+          Seq(RhoType.String("sender"), (RhoType.ByteArray(pk)), ackCh)
```
Comment (JosephDenman) [inline casper/src/test/scala/coop/rchain/casper/helper/BlockDataContract.scala]: Nit: extra parenthesis

File: casper/src/test/resources/PoSTest.rho [hunk 19]
```diff
@@ -463,6 +465,42 @@ match (
                 }
               }
             }
+          } |
+
+          contract getPosBalance(retCh) = {
```
Comment (JosephDenman) [inline casper/src/test/resources/PoSTest.rho:475]: This helper contract should be introduced in the next commit where it's used.

File: casper/src/main/resources/PoS.rhox [hunk 20]
```diff
@@ -344,17 +405,29 @@ new  PoS,
...
+        contract deposit (@fromUserPk, @amount, @activeValidators, returnCh) = {
           new vaultCh,
               revAddressCh,
+              indexSender,
+              getBlockData(`rho:block:data`), blockDataCh,
               authKeyCh in {
             @RevVault!("deployerAuthKey", *authKeyCh) |
             for (@authKey <- authKeyCh) {
               revAddressOps!("fromPublicKey", fromUserPk, *revAddressCh) |
+              getBlockData!(*blockDataCh) |
```
Comment (JosephDenman) [inline casper/src/main/resources/PoS.rhox:422]: Again, this would be more readable if the produce on `getBlockData` was in the same scope as the consume.

File: casper/src/main/resources/PoS.rhox (hunk 12 above)
Comment (JosephDenman) [reply casper/src/main/resources/PoS.rhox:418]: This is still the case after your changes. I can't bear Rholang being harder to read than it already is.

File: casper/src/test/resources/PoSTest.rho (hunk 17 above)
Comment (JosephDenman) [reply casper/src/test/resources/PoSTest.rho:494]: @odeac The PoS vault will always appear empty from the outside. It's only used as a temporary source vault for transfers during `closeBlock`, so I don't think it should make it to the API.

File: casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala [hunk 21]
```diff
@@ -15,7 +15,7 @@ class PoSSpec
     extends RhoSpec(
       CompiledRholangSource("PoSTest.rho", NormalizerEnv.Empty),
       Seq.empty,
-      120.seconds,
+      240.seconds,
```
Comment (JosephDenman) [inline casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala]: What did you determine was the source of this increase?

File: casper/src/main/resources/PoS.rhox (hunk 16 above)
Comment (JosephDenman) [reply casper/src/main/resources/PoS.rhox:108]: I agree, but I don't see  another way to access the vaults in Rholang. One hack would be to interpolate unforgeable names into the Rholang source code (for the necessary methods) and call those names through system processes in the tests. We might have to do that with some of the current public methods in `PoS` that will need to be private in production. If I'm not mistaken, `closeBlock` is an example.

**Comment by JosephDenman** [comment]
@tymm It's getting there. The main issue for me is the `getValidatorVaults` method that is introduced solely for testing. As I mentioned, interpolating an unforgeable name into `PoS.rhox` to represent a `getValidatorVaults` method might be a feasible approach to getting a handle on the vaults for testing. We'll have to solve the issue of protocol private methods (like `closeBlock`) in Rholang system contracts regardless.

Also, I think that `take` should be introduced as a method. It's not a higher-order function and implementing it as a built-in is likely to be far more performant that implementing it in `ListOps`.

File: casper/src/main/resources/PoS.rhox (hunk 5 above)
Comment (tymm) [reply casper/src/main/resources/PoS.rhox:515]: There can be more available validators than active ones, right? If that's the case the answer is "no".

File: casper/src/main/resources/PoS.rhox (hunk 12 above)
Comment (tymm) [reply casper/src/main/resources/PoS.rhox:418]: Sorry, I was sure that I had changed that already.

File: casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala (hunk 21 above)
Comment (tymm) [reply casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala]: `ListOps!take`.

File: casper/src/test/resources/PoSTest.rho (hunk 15 above)
Comment (tymm) [reply casper/src/test/resources/PoSTest.rho]: It's confusing and I removed it.
I guess what I meant is that once `closeBlock` is done, we are good to set the block sender to one of the active validators.

File: casper/src/test/resources/PoSTest.rho (hunk 17 above)
Comment (tymm) [reply casper/src/test/resources/PoSTest.rho:494]: @odeac we can make it part of the PoS API once we need it there.

File: casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala (hunk 21 above)
Comment (JosephDenman) [reply casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala]: I guess that's another reason to make `take` a built-in.

File: casper/src/main/resources/PoS.rhox (hunk 5 above)
Comment (odeac) [reply casper/src/main/resources/PoS.rhox:515]: but the take is not fair so it's not correct

Comment (tymm) [reply casper/src/main/resources/PoS.rhox:515]: Yeah, I put a comment about that.
Once we have on-chain randomness we can shuffle before we take.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1430]: @JosDenmark 's argument that "take" is not a higher-order function so it should be a method sounds fair to me

File: casper/src/test/resources/PoSTest.rho (hunk 17 above)
Comment (odeac) [reply casper/src/test/resources/PoSTest.rho:494]: ok

File: casper/src/main/resources/PoS.rhox (hunk 16 above)
Comment (odeac) [reply casper/src/main/resources/PoS.rhox:108]: We could extract the "vault per validator" logic into a separate vault-like contract that would have an address and a `balance` method, the same as all the "normal" vaults. The PoS contract will not really care about the actual implementation of its vault and the vault-per-validator logic can be tested outside of PoS tests.

**Comment by 9rb** [comment]
replaced by 2795. Some of the reviewer conversation may still be relevant.
